In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [3]:
data_sets_names = ['inventories', 'inventory_sets', 'sets', 'themes', 'inventory_minifigs', 'minifigs', 'inventory_parts',
             'part_categories', 'parts', 'colors', 'part_relationships', 'elements']

data_sets = [inventories, inventory_sets, sets, themes, inventory_minifigs, minifigs, inventory_parts,
             part_categories, parts, colors, part_relationships, elements]

# iterating over each data set and displaying respectives head info and describe
for i, val in enumerate(data_sets_names):
    print(f'\n{val} dataset')
    display(data_sets[i].head())
    display(data_sets[i].info())
    display(data_sets[i].describe())
    

### ENTER CODE HERE ###


inventories dataset


id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31587 entries, 0 to 31586
Data columns (total 3 columns):
id         31587 non-null int64
version    31587 non-null int64
set_num    31587 non-null object
dtypes: int64(2), object(1)
memory usage: 740.4+ KB


None

id       version
count   31587.000000  31587.000000
mean    43538.531389      1.077405
std     32458.542329      0.515993
min         1.000000      1.000000
25%     12193.500000      1.000000
50%     51076.000000      1.000000
75%     62254.500000      1.000000
max    122472.000000     11.000000


inventory_sets dataset


inventory_id  set_num  quantity
0            35  75911-1         1
1            35  75912-1         1
2            39  75048-1         1
3            39  75053-1         1
4            50   4515-1         1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3413 entries, 0 to 3412
Data columns (total 3 columns):
inventory_id    3413 non-null int64
set_num         3413 non-null object
quantity        3413 non-null int64
dtypes: int64(2), object(1)
memory usage: 80.1+ KB


None

inventory_id     quantity
count    3413.000000  3413.000000
mean    25839.150601     1.857017
std     31384.971888     6.123060
min        35.000000     1.000000
25%      6117.000000     1.000000
50%     12465.000000     1.000000
75%     27003.000000     1.000000
max    122081.000000    60.000000


sets dataset


set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1979        67         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18576 entries, 0 to 18575
Data columns (total 5 columns):
set_num      18576 non-null object
name         18576 non-null object
year         18576 non-null int64
theme_id     18576 non-null int64
num_parts    18576 non-null int64
dtypes: int64(3), object(2)
memory usage: 725.8+ KB


None

year      theme_id     num_parts
count  18576.000000  18576.000000  18576.000000
mean    2006.200581    405.924365    160.666344
std       13.749798    197.809058    395.885329
min     1949.000000      1.000000      0.000000
25%     2000.000000    246.000000      6.000000
50%     2010.000000    473.000000     37.000000
75%     2017.000000    525.000000    147.000000
max     2022.000000    719.000000  11695.000000


themes dataset


id            name  parent_id
0   1         Technic        NaN
1   3     Competition        1.0
2   4  Expert Builder        1.0
3  16      RoboRiders        1.0
4  17  Speed Slammers        1.0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 444 entries, 0 to 443
Data columns (total 3 columns):
id           444 non-null int64
name         444 non-null object
parent_id    306 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 10.5+ KB


None

id   parent_id
count  444.000000  306.000000
mean   410.659910  346.516340
std    208.165803  198.728717
min      1.000000    1.000000
25%    229.750000  186.000000
50%    438.500000  409.500000
75%    597.250000  507.000000
max    719.000000  697.000000


inventory_minifigs dataset


inventory_id     fig_num  quantity
0             3  fig-001549         1
1             4  fig-000764         1
2            19  fig-000555         1
3            25  fig-000574         1
4            26  fig-000842         1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18144 entries, 0 to 18143
Data columns (total 3 columns):
inventory_id    18144 non-null int64
fig_num         18144 non-null object
quantity        18144 non-null int64
dtypes: int64(2), object(1)
memory usage: 425.4+ KB


None

inventory_id      quantity
count   18144.000000  18144.000000
mean    26470.198302      1.066138
std     30578.235290      0.831258
min         3.000000      1.000000
25%      6901.000000      1.000000
50%     13832.000000      1.000000
75%     32633.000000      1.000000
max    122344.000000    100.000000


minifigs dataset


fig_num                             name  num_parts
0  fig-000001               Toy Store Employee          4
1  fig-000002                     Customer Kid          4
2  fig-000003            Assassin Droid, White          8
3  fig-000004                     Basic Figure          4
4  fig-000005  Captain America with Short Legs          3

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11827 entries, 0 to 11826
Data columns (total 3 columns):
fig_num      11827 non-null object
name         11827 non-null object
num_parts    11827 non-null int64
dtypes: int64(1), object(2)
memory usage: 277.3+ KB


None

num_parts
count  11827.000000
mean       4.847637
std        3.562961
min        0.000000
25%        4.000000
50%        4.000000
75%        5.000000
max      148.000000


inventory_parts dataset


inventory_id        part_num  color_id  quantity is_spare
0             1        48379c01        72         1        f
1             1           48395         7         1        f
2             1  stickerupn0077      9999         1        f
3             1         upn0342         0         1        f
4             1         upn0350        25         1        f

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 989529 entries, 0 to 989528
Data columns (total 5 columns):
inventory_id    989529 non-null int64
part_num        989529 non-null object
color_id        989529 non-null int64
quantity        989529 non-null int64
is_spare        989529 non-null object
dtypes: int64(3), object(2)
memory usage: 37.7+ MB


None

inventory_id       color_id       quantity
count  989529.000000  989529.000000  989529.000000
mean    30715.103387     124.593946       3.308733
std     32809.666367     854.794642       9.914533
min         1.000000      -1.000000       1.000000
25%      7928.000000       4.000000       1.000000
50%     15538.000000      15.000000       2.000000
75%     42590.000000      71.000000       3.000000
max    122472.000000    9999.000000    3064.000000


part_categories dataset


id                     name
0   1               Baseplates
1   3            Bricks Sloped
2   4  Duplo, Quatro and Primo
3   5           Bricks Special
4   6            Bricks Wedged

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
id      66 non-null int64
name    66 non-null object
dtypes: int64(1), object(1)
memory usage: 1.2+ KB


None

id
count  66.000000
mean   35.363636
std    19.411049
min     1.000000
25%    19.250000
50%    35.500000
75%    51.750000
max    68.000000


parts dataset


part_num                                        name  part_cat_id  \
0   003381                 Sticker Sheet for Set 663-1           58   
1   003383         Sticker Sheet for Sets 618-1, 628-2           58   
2   003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3   003429                Sticker Sheet for Set 1550-1           58   
4   003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   

  part_material  
0       Plastic  
1       Plastic  
2       Plastic  
3       Plastic  
4       Plastic

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44781 entries, 0 to 44780
Data columns (total 4 columns):
part_num         44781 non-null object
name             44781 non-null object
part_cat_id      44781 non-null int64
part_material    44781 non-null object
dtypes: int64(1), object(3)
memory usage: 1.4+ MB


None

part_cat_id
count  44781.000000
mean      39.541971
std       21.844951
min        1.000000
25%       18.000000
50%       42.000000
75%       60.000000
max       68.000000


colors dataset


id            name     rgb is_trans
0  -1       [Unknown]  0033B2        f
1   0           Black  05131D        f
2   1            Blue  0055BF        f
3   2           Green  237841        f
4   3  Dark Turquoise  008F9B        f

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 4 columns):
id          190 non-null int64
name        190 non-null object
rgb         190 non-null object
is_trans    190 non-null object
dtypes: int64(1), object(3)
memory usage: 6.1+ KB


None

id
count   190.000000
mean    479.394737
std     820.735479
min      -1.000000
25%      61.250000
50%     182.500000
75%    1015.750000
max    9999.000000


part_relationships dataset


rel_type child_part_num parent_part_num
0        P    87079pr9974           87079
1        P     3960pr9971            3960
2        R    98653pr0003     98086pr0003
3        R    98653pr0003    98088pat0003
4        R    98653pr0003    98089pat0003

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24508 entries, 0 to 24507
Data columns (total 3 columns):
rel_type           24508 non-null object
child_part_num     24508 non-null object
parent_part_num    24508 non-null object
dtypes: object(3)
memory usage: 574.5+ KB


None

rel_type child_part_num parent_part_num
count     24508          24508           24508
unique        6          22339            3858
top           P           3814           3626c
freq      19745             23            2353


elements dataset


element_id  part_num  color_id
0    6300211  67906c01        14
1    4566309      2564         0
2    4275423     53657      1004
3    6194308     92926        71
4    6229123     26561         4

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62684 entries, 0 to 62683
Data columns (total 3 columns):
element_id    62684 non-null object
part_num      62684 non-null object
color_id      62684 non-null int64
dtypes: int64(1), object(2)
memory usage: 1.4+ MB


None

color_id
count  62684.000000
mean     608.309983
std     2210.989312
min       -1.000000
25%       14.000000
50%       30.000000
75%      148.000000
max     9999.000000

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [4]:
# using shape function we can determine the number of rows in a given data set
def data_count(data):
    return int(data.shape[0])
### ENTER CODE HERE ###

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [5]:
data_count(sets)
### CODE CHECK ###

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [6]:
# using num_parts coulumn of sets table we can determine average number of parts (round() is used to rounded of result to nearest integer)
Q2 = round(sets['num_parts'].mean())
Q2
### ENTER CODE HERE ###

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [7]:
# using num_parts coulumn of sets table we can determine median number of parts (round() is used to rounded of result to nearest integer)
Q3 = round(sets['num_parts'].median())
Q3
### ENTER CODE HERE ###

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [8]:
# using max function on num_parts 
Q4 = sets['num_parts'].max()
Q4
### ENTER CODE HERE ###

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [9]:
# using filter to find Q5
Q5 = sets[sets['num_parts'] == Q4]
Q5
### ENTER CODE HERE ###

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [10]:
# using filter on set_num to find all rows for a particular set_num
def select_set_row(set_num):
    return sets[sets['set_num' == set_num]]
### ENTER CODE HERE ###

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [11]:
def select_set_numparts(number_of_parts):
    # filter sets_data using given num_parts and then sort the result by year and then set_num
    filtered_data = sets[sets['num_parts'] == number_of_parts]
    
    # sort dataframe by year and then set_num
    sorted_df = filtered_data.sort_values(by = ['year', 'set_num'])
    
    return sorted_df
### ENTER CODE HERE ###

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [12]:
select_set_numparts(0)
### CODE CHECK ###

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [13]:
# using year coulumn of sets table we can determine oldest set year
Q8A = int(sets.year.min())
display(Q8A)
### ENTER CODE HERE ###

1949

In [14]:
# using year coulumn of sets table we can determine newest set year
Q8B = int(sets.year.max())
display(Q8B)
### ENTER CODE HERE ###

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [15]:
def select_set_year(year):
    # filter sets_data using given year and then sort the result by set_num
    filtered_data = sets[sets['year'] == year]
    
    # sort dataframe by set_num
    sorted_df = filtered_data.sort_values(by = ['set_num'])
    
    return sorted_df
### ENTER CODE HERE ###

In [16]:
Q9A = select_set_year(Q8A)
display(Q9A)

Q9B = select_set_year(Q8B)
display(Q9B)
### ENTER CODE HERE ###

set_num                               name  year  theme_id  num_parts
11057  700.1-1         Extra-Large Gift Set (ABB)  1949       365        142
11070  700.2-1               Large Gift Set (ABB)  1949       365        178
11075  700.3-1              Medium Gift Set (ABB)  1949       365        142
11080  700.A-1              Small Brick Set (ABB)  1949       371         24
11081  700.B-1  Small Doors and Windows Set (ABB)  1949       371         12

set_num                                               name  \
336            10297-1                                     Boutique Hotel   
600            10781-1                          Spider-Man’s Techno Trike   
601            10782-1                      Hulk vs. Rhino Truck Showdown   
602            10783-1                        Spider-Man at Doc Ock’s Lab   
603            10784-1                     Spider-Man Webquarters Hangout   
...                ...                                                ...   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   
17207  9780794449247-1  DC Comics Super Heroes: Gotham City’s New Defe...   
17726    DOTS-SERIES-6                    DOTS Collectible Tiles Series 6   

       year  theme_id  num_parts  
336    2022       155       3068  
600    2022       706         59  
601    2022       706        110  
602    2022       706        131  
603    2022       706        155  
...     ...       ...        ...  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  
17207  2022       497          7  
17726  2022       688         16  

[183 rows x 5 columns]

**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [17]:
# using year coulumn of sets table we can determine median year

Q10 = int(sets.year.median())
display(Q10)
### ENTER CODE HERE ###

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [18]:
# using group by on year we can determine number of sets for each year and calculating mean over it to find number of sets per year
Q11 = int(sets.groupby('year').size().mean())

display(Q11)
### ENTER CODE HERE ###

258

**Q12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [19]:
# merging sets and names datasets on theme_id and id respectively and renaming name cloumns after merging.
merged_sets_themes = pd.merge(sets, themes, left_on='theme_id', right_on='id')
merged_sets_themes.rename(columns = {'name_x': 'name_sets', 'name_y': 'name_themes'}, inplace = True)

merged_sets_themes.head()

set_num                       name_sets  year  theme_id  num_parts  id  \
0   001-1                           Gears  1965         1         43   1   
1   002-1  4.5V Samsonite Gears Motor Set  1965         1          3   1   
2  1030-1  TECHNIC I: Simple Machines Set  1985         1        211   1   
3  1038-1              ERBIE the Robo-Car  1985         1        120   1   
4  1039-1            Manual Control Set 1  1986         1         39   1   

  name_themes  parent_id  
0     Technic        NaN  
1     Technic        NaN  
2     Technic        NaN  
3     Technic        NaN  
4     Technic        NaN

In [20]:
def theme_by_year(year):
    # filter data from merged_sets_themes based on given year and then sorted by id
    sorted_df = merged_sets_themes[merged_sets_themes['year'] == year].sort_values(by = ['id']) 
    
    # drop duplicates based on theme id
    sorted_df = sorted_df.drop_duplicates(subset='id', keep="first")
    
    # drop extra columns which are not required
    sorted_df.drop(columns = ['set_num', 'name_sets', 'theme_id', 'num_parts', 'parent_id', 'year'], inplace = True)
    
    # display the with required indexing
    display(sorted_df.reset_index(drop = True))
### ENTER CODE HERE ###

In [21]:
theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [22]:
theme_by_year(1996)
### CODE CHECK / OPTIONAL PRACTICE ###

id             name_themes
0     1                 Technic
1    21  Universal Building Set
2    67            Classic Town
3    90                Paradisa
4   126                   Space
5   131              Exploriens
6   142                 Spyrius
7   148               Pirates I
8   149         Imperial Armada
9   191             Dark Forest
10  201           Royal Knights
11  236                      9V
12  259                     NXT
13  278              Model Team
14  307                Aquazone
15  308               Aquanauts
16  311              Aquasharks
17  318                Belville
18  363                    Boat
19  399               Freestyle
20  443           Service Packs
21  453                 Technic
22  456                   Train
23  464           Time Cruisers
24  466                 Airport
25  468             Basic Model
26  469               Basic Set
27  471                 Ferries
28  473            Supplemental
29  475                 Western
30  503               Key Chain
31  504                   Duplo
32  505               Basic Set
33  507   Educational and Dacta
34  516       Duplo and Explore
35  525                 Technic
36  532            Supplemental
37  533                    Town
38  534  Universal Building Set
39  632                    Town
40  634                  Trains
41  652               Playhouse
42  664                   Toolo

**Q13:** What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).

In [23]:
# grouping themes based on ids and finding size associated with each id
size_for_each_id = merged_sets_themes.groupby('id').size().reset_index()

# find max size from above results
temp = size_for_each_id[size_for_each_id[0] == size_for_each_id[0].max()].reset_index()

# find name of theme assoicated with max sets
Q13 = themes[themes['id'] == temp.iloc[0]['id']].reset_index().iloc[0]['name']

display(Q13)
### ENTER CODE HERE ###

'Gear'

**Q14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [24]:
def theme_by_name(theme_name):
    # initialize list of column names to be displayed
    cols_to_display = ['set_num', 'name_sets', 'year', 'num_parts', 'theme_id', 'name_themes']
    
    # display result with filter on name_themes, sorted by set_num and indexed from zero
    display(merged_sets_themes[merged_sets_themes['name_themes'] == theme_name].sort_values(by = ['set_num'])[cols_to_display].reset_index(drop=True))
### ENTER CODE HERE ###

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q13`.  You should return 1,904 sets that match that theme name.

In [25]:
theme_by_name(Q13)
### CODE CHECK ###

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [26]:
# using filter on merged_sets_themes we can determine the result
def theme_by_setnum(set_number):
    return merged_sets_themes[merged_sets_themes['set_num'] == set_number].iloc[0]['name_themes']
### ENTER CODE HERE ###

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [27]:
theme_by_setnum('WATCH-3')
### CODE CHECK ###

'Gear'

**Q16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q16`.  Instead you will use this information for later questions.

In [28]:
# merging multiple data frames (minifigs, inventory_minifigs, inventories, sets and themes to find required results
# 1. merging minifigs and inventory_minifigs on fig_num
# 2. merging merged_minifig_iminifig and inbentories left on inentory id and right on id
# 3. merging merged_minifig_iminifig_inventory with merged_sets_themes on set_num

merged_minifig_iminifig = pd.merge(minifigs, inventory_minifigs, on='fig_num')
merged_minifig_iminifig_inventory = pd.merge(merged_minifig_iminifig, inventories, left_on='inventory_id', right_on='id')
merged_minifig_iminifig_inventory_sets = pd.merge(merged_minifig_iminifig_inventory, merged_sets_themes, on='set_num')

# renaming columns after merging
merged_minifig_iminifig_inventory_sets.rename(columns = {'num_parts_x': 'minifigs_num_parts', 'num_parts_y': 'parts_num_parts',
                                                         'id_x': 'id_from_inventories', 'id_y': 'id_from_sets', 'name': 'fig_name'}, inplace = True)

# display final mrged results
merged_minifig_iminifig_inventory_sets.head()

fig_num                                           fig_name  \
0  fig-000001                                 Toy Store Employee   
1  fig-000002                                       Customer Kid   
2  fig-000003                              Assassin Droid, White   
3  fig-000124                           Chewbacca, Reddish Brown   
4  fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...   

   minifigs_num_parts  inventory_id  quantity  id_from_inventories  version  \
0                   4         42484         1                42484        1   
1                   4         42484         1                42484        1   
2                   8         15042         1                15042        1   
3                   3         15042         1                15042        1   
4                   5         15042         1                15042        1   

   set_num            name_sets  year  theme_id  parts_num_parts  \
0  31105-1  Townhouse Toy Store  2020       672              554   
1  31105-1  Townhouse Toy Store  2020       672              554   
2  10188-1           Death Star  2008       171             3809   
3  10188-1           Death Star  2008       171             3809   
4  10188-1           Death Star  2008       171             3809   

   id_from_sets                name_themes  parent_id  
0           672             Creator 3-in-1       22.0  
1           672             Creator 3-in-1       22.0  
2           171  Ultimate Collector Series      158.0  
3           171  Ultimate Collector Series      158.0  
4           171  Ultimate Collector Series      158.0

In [29]:
# filter data as per Q16 requirement
filtered = merged_minifig_iminifig_inventory_sets[merged_minifig_iminifig_inventory_sets['name_sets'] == 'Millennium Falcon']

# filter max number of parts result using max on num_parts
max_entry = filtered[filtered['parts_num_parts'] == filtered.parts_num_parts.max()]

display(max_entry)
### ENTER CODE HERE ###

fig_num                                           fig_name  \
5946  fig-001714                              Chewbacca, Dark Brown   
5947  fig-001810                             Finn, Black Undershirt   
5948  fig-002057                                 Rey, Dark Tan Robe   
5949  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
5950  fig-002544                                               BB-8   
5951  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
5952  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
5953  fig-002551                               Han Solo, Old, Angry   

      minifigs_num_parts  inventory_id  quantity  id_from_inventories  \
5946                   3         19670         1                19670   
5947                   4         19670         1                19670   
5948                   4         19670         1                19670   
5949                   3         19670         1                19670   
5950                   2         19670         1                19670   
5951                   4         19670         1                19670   
5952                   4         19670         1                19670   
5953                   4         19670         1                19670   

      version  set_num          name_sets  year  theme_id  parts_num_parts  \
5946        1  75192-1  Millennium Falcon  2017       171             7541   
5947        1  75192-1  Millennium Falcon  2017       171             7541   
5948        1  75192-1  Millennium Falcon  2017       171             7541   
5949        1  75192-1  Millennium Falcon  2017       171             7541   
5950        1  75192-1  Millennium Falcon  2017       171             7541   
5951        1  75192-1  Millennium Falcon  2017       171             7541   
5952        1  75192-1  Millennium Falcon  2017       171             7541   
5953        1  75192-1  Millennium Falcon  2017       171             7541   

      id_from_sets                name_themes  parent_id  
5946           171  Ultimate Collector Series      158.0  
5947           171  Ultimate Collector Series      158.0  
5948           171  Ultimate Collector Series      158.0  
5949           171  Ultimate Collector Series      158.0  
5950           171  Ultimate Collector Series      158.0  
5951           171  Ultimate Collector Series      158.0  
5952           171  Ultimate Collector Series      158.0  
5953           171  Ultimate Collector Series      158.0

**Q17:** 
- How many minifigs are contained in the set from `Q16`?  Save this as `Q17` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [30]:
# using minifgs num_part_column of max_entry datafram, we can determine number of minifigs contained
Q17 = int(max_entry.minifigs_num_parts.sum())

display(Q17)
### ENTER CODE HERE ###

28

**Q18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [31]:
def get_minifigs(set_number, version_number):
    # initialize list of column names to be displayed
    cols_to_display = ['set_num', 'fig_num', 'name', 'num_parts']
    
    # filter merged_minifig_iminifig_inventory data frame for given set number and version number and display the required columns with indexing from 0    
    filtered_df = merged_minifig_iminifig_inventory[(merged_minifig_iminifig_inventory['set_num'] == set_number) &
                                                    (merged_minifig_iminifig_inventory['version'] == version_number)].reset_index()[cols_to_display]
    
    filtered_df.rename(columns = {'name': 'fig_name', 'num_parts': 'fig_num_parts'}, inplace = True)
    
    display(filtered_df)
### ENTER CODE HERE ###

**Q19:** Using the above function, output information about the minifigs from the set discussed in `Q16`.  Save this output as `Q19`.

In [32]:
get_minifigs('75192-1', 1)
### ENTER CODE HERE ###

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [33]:
get_minifigs('10217-1', 1)
get_minifigs('10217-1', 2)

### CODE CHECK ###

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [34]:
def minifigs_from_themes(theme_id):
    # filter merged_minifig_iminifig_inventory_sets for a given theme_id and extract required columns
    filtered_df = merged_minifig_iminifig_inventory_sets[merged_minifig_iminifig_inventory_sets['theme_id'] == theme_id][['fig_num', 'fig_name']]
    
    # sort dataframe by fig_num
    filtered_df.sort_values(by = 'fig_num', inplace = True)
    
    # drop duplicates from the result
    filtered_df.drop_duplicates(inplace=True)
    
    display(filtered_df.reset_index(drop=True))
### ENTER CODE HERE ###

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q16`.  There should be 149 minifigs output for that respective theme.

In [35]:
minifigs_from_themes(171)
### CODE CHECK ###

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [36]:
def sets_from_minifig(minifig_number):
    # initialize list of column names to be displayed
    cols_to_display = ['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']
    
    # filter merged_minifig_iminifig_inventory_sets for a given minifig number and extract required columns
    filtered_df = merged_minifig_iminifig_inventory_sets[merged_minifig_iminifig_inventory_sets['fig_num'] == minifig_number][cols_to_display]
    
    # sort dataframe by fig_num and then version
    filtered_df.sort_values(by = ['fig_num', 'version'], inplace = True)
    
    return filtered_df.reset_index(drop = True)
### ENTER CODE HERE ###     

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [37]:
sets_from_minifig('fig-001039')
### CODE CHECK ###

set_num  version            name_sets     fig_num  \
0   3061-1        1       City Park Café  fig-001039   
1  30100-1        1  Andrea on the Beach  fig-001039   
2  30107-1        1       Birthday Party  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q22:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [38]:
# merging multiple data frames (minifigs, inventory_minifigs, inventories, sets and themes to find required results
# 1. merging merged_sets_themes with inventories on set_num
# 2. merging inventory_parts and parts on part_num
# 3. merging merged_iparts_parts and color left on color_id and right on id
# 4. merging merged_iparts_parts_colors with merged_sets_themes_inventories left on inventory_id and right on inventories_id

cols_to_display_for_set_parts = ['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color','quantity','is_spare', 'num_parts']

merged_sets_themes_inventories = pd.merge(merged_sets_themes, inventories, on = 'set_num')
merged_sets_themes_inventories.rename(columns = {'id_y': 'inventories_id'}, inplace = True)

merged_iparts_parts = pd.merge(inventory_parts[['inventory_id', 'part_num', 'color_id', 'quantity', 'is_spare']], parts[['part_num', 'name']], on='part_num')
merged_iparts_parts_colors = pd.merge(merged_iparts_parts, colors[['id', 'name']], left_on='color_id', right_on='id')

# rename and drop columns
merged_iparts_parts_colors.rename(columns = {'name_x': 'name_part', 'name_y':'name_color'}, inplace=True)
merged_iparts_parts_colors.drop(columns = ['id'], inplace=True)


merged_iparts_parts_colors_sti = pd.merge(merged_iparts_parts_colors, merged_sets_themes_inventories[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'inventories_id', 'num_parts']],
                                          left_on='inventory_id', right_on='inventories_id')[cols_to_display_for_set_parts]

merged_iparts_parts_colors_sti.head()

set_num  version                                          name_sets  year  \
0  7922-1        1  McDonald's Sports Set Number 6 - Orange Vest S...  2004   
1  7922-1        1  McDonald's Sports Set Number 6 - Orange Vest S...  2004   
2  7922-1        1  McDonald's Sports Set Number 6 - Orange Vest S...  2004   
3  7922-1        1  McDonald's Sports Set Number 6 - Orange Vest S...  2004   
4  7922-1        1  McDonald's Sports Set Number 6 - Orange Vest S...  2004   

     name_themes        part_num  \
0  Gravity Games        48379c01   
1  Gravity Games         upn0342   
2  Gravity Games           48395   
3  Gravity Games  stickerupn0077   
4  Gravity Games         upn0350   

                                           name_part            name_color  \
0  Large Figure Torso and Legs, Promo Figure Base...      Dark Bluish Gray   
1    Sports Promo Paddle from McDonald's Sports Sets                 Black   
2   Sports Snowboard from McDonald's Promotional Set            Light Gray   
3                       Sticker Sheet for Set 7922-1  [No Color/Any Color]   
4  Sports Promo Figure Head Torso Assembly McDona...                Orange   

   quantity is_spare  num_parts  
0         1        f          5  
1         1        f          5  
2         1        f          5  
3         1        f          5  
4         1        f          5

In [39]:
def set_parts(set_number, version_number):
    # filter merged_iparts_parts_colors_sti on given set number and version number
    filtered_df = merged_iparts_parts_colors_sti[(merged_iparts_parts_colors_sti['set_num'] == set_number) & (merged_iparts_parts_colors_sti['version'] == version_number)]
    
    # sort dataframe by part_num and then name_color
    filtered_df.sort_values(by = ['part_num', 'name_color'], inplace = True)
    
    # return final result with indexing from 0
    return filtered_df.reset_index(drop = True)
### ENTER CODE HERE ###

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q16`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [40]:
result = set_parts('75192-1', 1)

display(result)

print("Number of parts", result.part_num.nunique())

print("Number of spare parts", data_count(result[result.is_spare == 't']))

### CODE CHECK ###

C:\Users\M BILAL\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  num_parts  
0     Dark Bluish Gray         3        f       7541  
1    Light Bluish Gray        12        f       7541  
2     Dark Bluish Gray         5        f       7541  
3                White        22        f       7541  
4                  Tan         3        f       7541  
..                 ...       ...      ...        ...  
725  Light Bluish Gray        70        f       7541  
726              Black         6        f       7541  
727   Dark Bluish Gray        25        f       7541  
728  Light Bluish Gray         2        f       7541  
729  Light Bluish Gray         3        f       7541  

[730 rows x 11 columns]

Number of parts 376
Number of spare parts 54


  **Q23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q23`. 

In [41]:
Q23 = list(colors.name.unique())
Q23.sort()

display(Q23)
### ENTER CODE HERE ###

['Aqua',
 'Black',
 'Blue',
 'Blue-Violet',
 'Bright Green',
 'Bright Light Blue',
 'Bright Light Orange',
 'Bright Light Yellow',
 'Bright Pink',
 'Brown',
 'Chrome Antique Brass',
 'Chrome Black',
 'Chrome Blue',
 'Chrome Gold',
 'Chrome Green',
 'Chrome Pink',
 'Chrome Silver',
 'Copper',
 'Coral',
 'Dark Azure',
 'Dark Blue',
 'Dark Blue-Violet',
 'Dark Bluish Gray',
 'Dark Brown',
 'Dark Flesh',
 'Dark Gray',
 'Dark Green',
 'Dark Orange',
 'Dark Pink',
 'Dark Purple',
 'Dark Red',
 'Dark Tan',
 'Dark Turquoise',
 'Earth Orange',
 'Fabuland Brown',
 'Fabuland Orange',
 'Flat Dark Gold',
 'Flat Silver',
 'Flesh',
 'Glitter Trans-Clear',
 'Glitter Trans-Dark Pink',
 'Glitter Trans-Light Blue',
 'Glitter Trans-Neon Green',
 'Glitter Trans-Orange',
 'Glitter Trans-Purple',
 'Glow In Dark Opaque',
 'Glow In Dark Trans',
 'Glow in Dark White',
 'Green',
 'Lavender',
 'Light Aqua',
 'Light Blue',
 'Light Bluish Gray',
 'Light Flesh',
 'Light Gray',
 'Light Green',
 'Light Lime',
 'Light 

**Q24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q24` (as an integer).

In [42]:
# filter on colors is_trans column and counting rows with data_count function defined above
Q24 = data_count(colors[colors['is_trans'] == 't'])

display(Q24)
### ENTER CODE HERE ###

38

**Q25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [43]:
def get_part_colors(part_number):
    # filter merged_iparts_parts_colors on given part number and extract required columns
    filtered_df = merged_iparts_parts_colors[merged_iparts_parts_colors['part_num'] == part_number][['part_num', 'color_id', 'name_color']]
    
    # sort dataframe by color_id
    filtered_df.sort_values(by = ['color_id'], inplace = True)
    
    # return final result with indexing from 0
    return filtered_df.reset_index(drop = True)
### ENTER CODE HERE ###